### PR TITLE
[TIMOB-19163] Android: Removal of Apache HTTP Client

### PR DESCRIPTION
**Pull request by ashcoding**
-Implemented HttpURLConnection for Android M to use

### CODE REVIEW

**Comment by ashcoding** [comment]
Jira:- https://jira.appcelerator.org/browse/TIMOB-19163

Note: 
Method addAuthFactory has been removed in this PR 
http://appcelerator.github.io/appc-docs/latest/#!/api/Titanium.Network.HTTPClient-method-addAuthFactory

File: android/modules/network/src/java/ti/modules/titanium/network/CookieProxy.java [hunk 1]
```diff
@@ -41,29 +39,28 @@
...
-			setProperty(TiC.PROPERTY_SECURE, basicClientCookie.isSecure());
-			setProperty(TiC.PROPERTY_VERSION, basicClientCookie.getVersion());
+		if (cookie instanceof HttpCookie) {
+			httpCookie = (HttpCookie) cookie;
+			setProperty(TiC.PROPERTY_NAME, httpCookie.getName());
+			setProperty(TiC.PROPERTY_VALUE, httpCookie.getValue());
+			setProperty(TiC.PROPERTY_DOMAIN, httpCookie.getDomain());
+			// PROPERTY_EXPIRY_DATE not used instead, PROPERTY_MAX_AGE is used
+			// See http://developer.android.com/reference/java/net/HttpCookie.html for more info
+			setProperty(TiC.PROPERTY_MAX_AGE, httpCookie.getMaxAge());
```
Comment (hieupham007) [inline android/modules/network/src/java/ti/modules/titanium/network/CookieProxy.java]: Did you modify the doc for this replacement property?

File: android/modules/network/src/java/ti/modules/titanium/network/CookieProxy.java [hunk 2]
```diff
@@ -170,6 +162,6 @@ public String getName()
 	@Override
 	public String getApiName()
 	{
-		return "Ti.Network.Cookie";
+		return "Ti.Network.CookieProxy";
```
Comment (hieupham007) [inline android/modules/network/src/java/ti/modules/titanium/network/CookieProxy.java]: Any particular reason for this change? A typo?

File: android/modules/network/src/java/ti/modules/titanium/network/NetworkModule.java [hunk 3]
```diff
@@ -40,7 +40,7 @@
 public class NetworkModule extends KrollModule {
 
 	private static final String TAG = "TiNetwork";
-	private static CookieStore httpCookieStore;
+	private static java.net.CookieManager cookieManager;
```
Comment (hieupham007) [inline android/modules/network/src/java/ti/modules/titanium/network/NetworkModule.java]: Please import this so you can just use CookieManager

File: android/modules/network/src/java/ti/modules/titanium/network/NetworkModule.java [hunk 4]
```diff
@@ -299,13 +299,14 @@ public void onDestroy(Activity activity) {
...
 	}
-
-	public static CookieStore getHTTPCookieStoreInstance()
+	
+	public static java.net.CookieManager getCookieManagerInstance()
 	{
-		if (httpCookieStore == null) {
-			httpCookieStore = new BasicCookieStore();
+		if (cookieManager == null) {
+			cookieManager = new java.net.CookieManager();
```
Comment (hieupham007) [inline android/modules/network/src/java/ti/modules/titanium/network/NetworkModule.java]: same here

File: android/modules/network/src/java/ti/modules/titanium/network/TiAuthenticator.java [hunk 5]
```diff
@@ -0,0 +1,33 @@
+package ti.modules.titanium.network;
+
+import java.net.Authenticator;
+import java.net.PasswordAuthentication;
+
+public class TiAuthenticator extends Authenticator {
+	
+	String domain;
```
Comment (hieupham007) [inline android/modules/network/src/java/ti/modules/titanium/network/TiAuthenticator.java]: indentation

File: android/modules/network/src/java/ti/modules/titanium/network/TiHTTPClient.java [hunk 6]
```diff
@@ -168,327 +129,208 @@
...
+		connected = true;	
 
-			if (response == null) {
-				throw new IllegalArgumentException("HTTP response may not be null");
+		long contentLength;
+		
+		if (connection != null) {
+			TiHTTPClient c = this;
+			
+	        contentLength = connection.getContentLength();
```
Comment (hieupham007) [inline android/modules/network/src/java/ti/modules/titanium/network/TiHTTPClient.java]: indentations. Please auto indent each file you change.

File: android/modules/network/src/java/ti/modules/titanium/network/TiHTTPClient.java [hunk 7]
```diff
@@ -168,327 +129,208 @@
...
-						is = new GZIPInputStream(entity.getContent());
-					} else {
-						is = entity.getContent();
-					}
-					charset = EntityUtils.getContentCharSet(entity);
-				} else {
-					is = null;
 				}
+				//entity.consumeContent()
+				// TODO: Close inputstream
```
Comment (hieupham007) [inline android/modules/network/src/java/ti/modules/titanium/network/TiHTTPClient.java]: ?

File: android/modules/network/src/java/ti/modules/titanium/network/CookieProxy.java (hunk 2 above)
Comment (ashcoding) [inline android/modules/network/src/java/ti/modules/titanium/network/CookieProxy.java]: A Typo. A very bad typo. Thanks for catching this.

File: android/modules/network/src/java/ti/modules/titanium/network/NetworkModule.java (hunk 3 above)
Comment (ashcoding) [inline android/modules/network/src/java/ti/modules/titanium/network/NetworkModule.java]: There's another class with the same name that is imported:-
```java import android.webkit.CookieManager; ```
I don't think I can import it.

File: android/modules/network/src/java/ti/modules/titanium/network/NetworkModule.java (hunk 4 above)
Comment (ashcoding) [inline android/modules/network/src/java/ti/modules/titanium/network/NetworkModule.java]: same as above

File: android/modules/network/src/java/ti/modules/titanium/network/TiAuthenticator.java (hunk 5 above)
Comment (ashcoding) [inline android/modules/network/src/java/ti/modules/titanium/network/TiAuthenticator.java]: Will do

File: android/modules/network/src/java/ti/modules/titanium/network/TiHTTPClient.java (hunk 6 above)
Comment (ashcoding) [inline android/modules/network/src/java/ti/modules/titanium/network/TiHTTPClient.java]: will do.

File: android/modules/network/src/java/ti/modules/titanium/network/TiHTTPClient.java (hunk 7 above)
Comment (ashcoding) [inline android/modules/network/src/java/ti/modules/titanium/network/TiHTTPClient.java]: Got it. Will remove.

File: android/modules/network/src/java/ti/modules/titanium/network/CookieProxy.java (hunk 1 above)
Comment (ashcoding) [inline android/modules/network/src/java/ti/modules/titanium/network/CookieProxy.java]: Will modify the docs.

**Comment by hieupham007** [comment]
Ash, excellent work for this PR! Initial code review looks good, just one thing: I don't see any removal of Apache libraries (the jar itself). We should probably do that, along with update build scripts to remove any apache related libraries.

**Comment by hieupham007** [comment]
I've ran your test app, and ran into a few problems:
1. Test app have some layout issues when running on Galaxy Nexus (buttons overlap each other). Probably a good idea to put them in a ListView.
2. Your test suite is quite basic, thus doesn't cover advanced cases, or regressions. Depends on what's changed, it's probably a good idea to find existing tickets that are solved in areas impacted by this PR and test them. For instance, regarding cookies, these two tickets could serve as additional test cases:
https://jira.appcelerator.org/browse/TIMOB-14610
https://jira.appcelerator.org/browse/TIMOB-15434
A PR of this magnitude should have at least 10-15 tickets to test for regression.
3. Individual test cases don't have expected results or description (i.e what should the tester expects when running the test case). Currently all test cases result in an Alert dialog, most with wall of texts. This may hinder QE from verifying expected results.

**Comment by ashcoding** [comment]
Okay. I'll update the tests.

**Comment by hieupham007** [comment]
Regression: https://jira.appcelerator.org/browse/TIMOB-15380: ticket works fine in 4.1.0.GA (location is redirected properly)

**Comment by ashcoding** [comment]
@hieupham007 Will work on those 2 working. Thanks for finding those!

**Comment by ashcoding** [comment]
Fixed Regressions for:
1. https://jira.appcelerator.org/browse/TIMOB-15380: ticket works fine in 4.1.0.GA (location is redirected properly)
2. https://jira.appcelerator.org/browse/TIMOB-17585: 4.1.0.GA works fine, crashes with this PR

Check JIRA for updated test case for these 2.

**Comment by hieupham007** [comment]
Code reviewed and tested with KS, the test case in the ticket, and these tickets:
https://jira.appcelerator.org/browse/TIMOB-15244
https://jira.appcelerator.org/browse/TIMOB-15380
https://jira.appcelerator.org/browse/TIMOB-6411
https://jira.appcelerator.org/browse/TIMOB-17585

Request **accepted**.

**Comment by ashcoding** [comment]
Thank you.

**Pull request merged**